### PR TITLE
Fix R Activation

### LINKED ISSUE #4795: Help with r-ada `depends_on(r_rpart)` but rpart seems to be in the R core.
I recently added a package for *ada* that `depends_on(r-rpart)`.  I just tried to activate it and it failed.  If I understand what's happening, `rpart` is part of the R core and I should not have added a dependency for it in the package.

Is that correct?  If so I'll update the package.

### PR DESCRIPTION
**Pull request by tmerrick1**
Here is the fix to remove the recommended functions from the R installation per #3031 

Fixes #4795.

### CODE REVIEW

**Comment by adamjstewart** [comment]
This is quite a few commits for a single line change. Is your develop branch polluted with other commits that aren't in upstream develop?

**Comment by tmerrick1** [comment]
Yes, how do I get rid of them?  They never seem to leave.

Tom

**Comment by adamjstewart** [comment]
@tmerrick1 That depends. Can you give me the output of:
```
$ git remote -v
```

**Comment by adamjstewart** [comment]
The other possibility is that you've been creating branches off of existing PR branches instead of off of develop.

**Comment by tmerrick1** [comment]
origin  https://github.com/LLNL/spack.git (fetch)
origin  https://github.com/LLNL/spack.git (push)
tmerrick        git@github.com:tmerrick1/spack.git (fetch)
tmerrick        git@github.com:tmerrick1/spack.git (push)

**Comment by adamjstewart** [comment]
@tmerrick1 Alright, so run the following commands:
```
$ git checkout develop   # make sure you are on the develop branch
$ git pull -f origin develop  # force pull LLNL develop, overwriting your own
$ git push -f tmerrick develop  # force push to your fork, overwriting your own
```
Then from now on, every new branch you create will only have a single commit in it. If you want to fix this PR, the easiest way would probably be to delete it locally, create a new branch with the same name, make the change, commit it, and force push it to your fork.

**Comment by tmerrick1** [comment]
Added the missing recommended package, r-spatial.

**Comment by adamjstewart** [comment]
@tmerrick1 At this point, I think the only thing we need to do to merge this PR is to make sure that this change doesn't break any existing R packages. We need to make sure that all 240 R packages in Spack aren't missing any dependencies on these recommended packages. You can either do this by hand by checking CRAN and making sure everything is listed as a dependency, or you can `spack install` every R package in a for-loop. Once you confirm that everything installs, try running `spack activate` on every R package to make sure there are no conflicting files. If both of those tests check out, I think we are safe to merge this change.

**Comment by tmerrick1** [comment]
Let me check it this out.  It looks like it added in some development files that I did not add with the 'git add' command

File: var/spack/repos/builtin/packages/mesa/package.py [hunk 1]
```diff
@@ -41,15 +41,15 @@ class Mesa(AutotoolsPackage):
...
-    depends_on('libxfixes')
+    depends_on('libpthread-stubs', type=('build', 'link', 'run'))
+    depends_on('libdrm', type=('build', 'link', 'run'))
+    depends_on('openssl', type=('build', 'link', 'run'))
+    depends_on('libxcb@1.9.3:', type=('build', 'link', 'run'))
+    depends_on('libxshmfence@1.1:', type=('build', 'link', 'run'))
+    depends_on('libx11', type=('build', 'link', 'run'))
+    depends_on('libxext', type=('build', 'link', 'run'))
+    depends_on('libxdamage', type=('build', 'link', 'run'))
+    depends_on('libxfixes', type=('build', 'link', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mesa/package.py]: I assume this change wasn't supposed to end up in this PR?

File: var/spack/repos/builtin/packages/r-bindrcpp/package.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    homepage = "https://github.com/krlmlr/bindrcpp"
+    url      = "https://cran.r-project.org/src/contrib/bindrcpp_0.2.tar.gz"
+    list_url = "https://cran.r-project.org/src/contrib/Archive/bindrcpp"
+
+    version('0.2', '2ed7f19fd9a12587f882d90060e7a343')
+
+    depends_on('r-rcpp', type=('build', 'run'))
+    depends_on('r-bindr', type=('build', 'run'))
+    depends_on('r-plogr', type=('build'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/r-bindrcpp/package.py]: I'm not sure if I've ever seen a build-only dependency in an R package before. Can you explain?

File: var/spack/repos/builtin/packages/r-bit/package.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    and writing to bit is approximately as fast as accessing standard logicals
+    - mostly due to R's time for memory allocation. The package allows to work
+    with pre-allocated memory for return values by calling .Call() directly:
+    when evaluating the speed of C-access with pre-allocated vector memory,
+    coping from bit to logical requires only 70% of the time for copying from
+    logical to logical; and copying from logical to bit comes at a performance
+    penalty of 150%. the package now contains further classes for representing
+    logical selections: 'bitwhich' for very skewed selections and 'ri' for
+    selecting ranges of values for chunked processing. All three index classes
+    can be used for subsetting 'ff' objects (ff-2.1-0 and higher)."""
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/r-bit/package.py]: This description is a bit long. Normally it should be 1-3 sentences. This description is printed when you run `spack info` and is added to the module file.

File: var/spack/repos/builtin/packages/r-abind/package.py [hunk 4]
```diff
@@ -32,6 +32,10 @@ class RAbind(RPackage):
...
 
     homepage = "https://cran.r-project.org/"
-    url      = "https://cran.r-project.org/src/contrib/abind_1.4-3.tar.gz"
+    url      = "https://cran.r-project.org/src/contrib/abind_1.4-5.tar.gz"
+    list_url = "https://cran.r-project.org/src/contrib/Archive/abind"
 
+    version('1.4-5', '136f981e1c4f618b64a87faaa7797c97')
     version('1.4-3', '10fcf80c677b991bf263d38be35a1fc5')
+
+    depends_on('r@1.5.0:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/r-abind/package.py]: `type=('build', 'run')`

File: var/spack/repos/builtin/packages/r-ape/package.py [hunk 5]
```diff
@@ -44,7 +44,11 @@ class RApe(RPackage):
 
     homepage = "http://ape-package.ird.fr/"
     url      = "https://cran.r-project.org/src/contrib/ape_4.1.tar.gz"
+    list_url = "https://cran.r-project.org/src/contrib/Archive/ape"
 
     version('4.1', 'a9ed416d6d172d4b9682556cf692d7c2')
 
     depends_on('r@3.2:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/r-ape/package.py]: `type=('build', 'run')`

File: var/spack/repos/builtin/packages/r-backports/package.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    The 'root' of a project is defined as a directory that matches a certain
+    criterion, e.g., it contains a certain regular file."""
+
+    homepage = "https://github.com/mllg/backports"
+    url      = "https://cran.r-project.org/src/contrib/backports_1.1.0.tar.gz"
+    list_url = "https://cran.r-project.org/src/contrib/Archive/backports"
+
+    version('1.1.0', 'b97a71b026fd7ede0e449be93d160c17')
+
+    depends_on('r@3.0.0:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/r-backports/package.py]: `type=('build', 'run')`

File: var/spack/repos/builtin/packages/r-base64enc/package.py [hunk 7]
```diff
@@ -31,5 +31,8 @@ class RBase64enc(RPackage):
 
     homepage = "http://www.rforge.net/base64enc"
     url      = "https://cran.r-project.org/src/contrib/base64enc_0.1-3.tar.gz"
+    list_url = "https://cran.r-project.org/src/contrib/Archive/base64enc"
 
     version('0.1-3', '0f476dacdd11a3e0ad56d13f5bc2f190')
+
+    depends_on('r@2.9.0:')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/r-base64enc/package.py]: etc.

File: var/spack/repos/builtin/packages/r-rpart/package.py [hunk 8]
```diff
@@ -30,7 +30,11 @@ class RRpart(RPackage):
     survival trees."""
 
     homepage = "https://cran.r-project.org/package=rpart"
+<<<<<<< Updated upstream
     url      = "https://cran.r-project.org/src/contrib/rpart_4.1-10.tar.gz"
+=======
+    url      = "https://cran.r-project.org/src/contrib/rpart_4.1-11.tar.gz"
+>>>>>>> Stashed changes
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/r-rpart/package.py]: This looks like a merge mistake.

File: var/spack/repos/builtin/packages/r-xml2/package.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    homepage = "https://github.com/hadley/xml2"
+    url      = "https://cran.r-project.org/src/contrib/xml2_1.1.1.tar.gz"
+    list_url = "https://cran.r-project.org/src/contrib/Archive/xml2"
+
+    version('1.1.1', '768f7edc39c4baab6b6b9e7c7ec79fee')
+
+    depends_on('r@3.1.0:')
+
+    depends_on('r-rcpp@0.11.4.6:', type=('build', 'run'))
+    depends_on('r-bh')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/r-xml2/package.py]: `type=('build', 'run')`

**Comment by tmerrick1** [comment]
Did we want to put the depends_on('r') on all of the ones that do not have version requirements so we can get rid of the depends_on('r'...) in the class install?  This way it will only have to process one depends_on('r'...) per item - speeding things up.  But it means that people need to remember to do it in the package install.

**Comment by adamjstewart** [comment]
No, we should keep the `depends_on('r')` in the `RPackage` base class. It makes things simpler. I don't think it will add that much additional processing time as all of the restrictions should be merged before it tries to concretize the spec. But I must admit I don't know much about the concretization process. I wish there was a way to speed up R concretization. But @tgamblin is working on a new concretizer that will hopefully be faster.

**Comment by tmerrick1** [comment]
I think that I have all of the type=(‘build’, ‘run’) added to the r depends_on statements.

I still do not like the way that ‘git rm’ works.  It looks to me that it will erase these lines from the LLNL spack mesa package.

The reason that the package has only a build is that it is c++ macros that are only needed during compilation.

Tom

**Comment by adamjstewart** [comment]
Yes, it does look like it's going to remove that file. Try [this](https://stackoverflow.com/a/38744134/5828163).

**Comment by tmerrick1** [comment]
I think the ‘git rm’ messed that up:

$ git checkout -- mesa/package.py
error: pathspec 'var/spack/repos/builtin/packages/mesa/package.py' did not match any file(s) known to git.

And ‘git checkout’ only works to the last commit.

I think that I have it so it will end up with the “official” mesa package instead of deleting it.  I could not figure out a way to just delete my changes to a committed pull.

Is there any way to ignore one of the changed files during merge?

**Comment by adamjstewart** [comment]
I think the easiest way to remove the changes to `mesa` would be to checkout a fresh copy of `develop`, copy the `mesa` package somewhere, switch back to `r3031fix`, and replace the file with the unchanged `mesa` package. Then commit it, and if the file disappears from the changed files list you know you're all good.

I really want to get this merged ASAP. 278 changed files is just asking for a merge conflict to arise later.

@becker33 Can you approve now that the `r-spatial` package has been added?

**Comment by adamjstewart** [comment]
> I really want to get this merged ASAP. 278 changed files is just asking for a merge conflict to arise later.

Speak of the devil, there is already a merge conflict with `r-rjava`.

**Comment by tmerrick1** [comment]
So I do the following:

git reset --hard /upstream/develop
git pull

erase the mesa file

git checkout develop
git merge upstream/develop
git push origin

now do the git checkout of my fix and add the mesa file back in?

Tom

**Comment by adamjstewart** [comment]
@tmerrick1 Try:
```
# Update your develop branch
git checkout develop
git pull origin develop
git push tmerrick develop

# Copy the mesa package
cp var/spack/repos/builtin/packages/mesa/package.py ~

# Go back to your branch
git checkout r3031fix

# Replace the file
mv ~/package.py var/spack/repos/builtin/packages/mesa/package.py

# Commit and push
```

**Comment by tmerrick1** [comment]
It looks like it went back in time. I think mesa is gone, but all of the updates for the build type seem to be missing also.

Tom

**Comment by alalazo** [comment]
It seems `r` packages follow a `url` with some structure. Did we ever try to code `url` and `list_url` as property of `RPackage` (overriding in the package only the few nasty cases that need it)?

**Comment by adamjstewart** [comment]
@tmerrick1 @6110 is going to re-add the `list_url`s to every R package. Perhaps after it is merged, you can rebase this PR, or create a new PR. It would be great to have R activate-able before SC17!

**Comment by tgamblin** [comment]
@lingnanyuan @fanne-stat @akthoma: since you all are doing a lot of R work these days, do you mind taking a look at this PR?  I'd be interested to know whether it would help you out, particularly now that #5415 is merged (so you could have multiple views with different sets of R packages).

**Comment by fanne-stat** [comment]
I don't know if I understand the issue correctly. But there is a problem that some packages in the recommended packages may not be available, e.g. `tcltk`. https://cran.r-project.org/web/packages/tcltk/index.html. It says "Package ‘tcltk’ was removed from the CRAN repository".

**Comment by akthoma** [comment]
@tgamblin: looks good to me! For me, personally, the biggest problem that I occasionally have with R packages is discrepancy between the url that I feed to `spack create` and the versions it returns to me. `list_url` seems to fix that problem when it occurs. For instance:
```
[las_thoma15@build spack]$ spack create --force https://cran.r-project.org/src/contrib/geiger_2.0.6.tar.gz
==> This looks like a URL for geiger
==> Warning: Spack will not check SSL certificates. You need to update your Python to enable certificate verification.
==> Found 18 versions of geiger:
  
  2.0.5     https://cran.r-project.org/src/contrib/Archive/geiger/geiger_2.0.5.tar.gz
  2.0.3     https://cran.r-project.org/src/contrib/Archive/geiger/geiger_2.0.3.tar.gz
  2.0.1     https://cran.r-project.org/src/contrib/Archive/geiger/geiger_2.0.1.tar.gz
  2.0       https://cran.r-project.org/src/contrib/Archive/geiger/geiger_2.0.tar.gz
  1.99-3.1  https://cran.r-project.org/src/contrib/Archive/geiger/geiger_1.99-3.1.tar.gz
  1.99-3    https://cran.r-project.org/src/contrib/Archive/geiger/geiger_1.99-3.tar.gz
  1.99-2    https://cran.r-project.org/src/contrib/Archive/geiger/geiger_1.99-2.tar.gz
  1.3-1     https://cran.r-project.org/src/contrib/Archive/geiger/geiger_1.3-1.tar.gz
  1.2-14    https://cran.r-project.org/src/contrib/Archive/geiger/geiger_1.2-14.tar.gz
  ...
  0.2-5     https://cran.r-project.org/src/contrib/Archive/geiger/geiger_0.2-5.tar.gz

==> How many would you like to checksum? (default is 1, q to abort) 
```
as you can see, the "latest" version that is returned to me is 2.0.5, and versions that are only found in the archive rather than the url that I provided with `spack create`. Like I said though, adding `list_url` has fixed this for me in the past.

I am not meaning to open a whole new issue on this PR (it looks like #6110 should've fixed this for me)! I just have not worked enough with issues in spack activation and base packages to know whether I would benefit from this PR other than the re-addition of `list_url` initially introduced in this PR.

**Comment by alalazo** [comment]
That's what I call having conflicts! @tmerrick1 Care to rebase?

**Comment by alalazo** [comment]
Closing as we having no reply since November. Feel free to reopen when / if you rebase.

**Pull request closed without merging**